spootify/src/common/utils/spotifyHttp.js

```diff
@@ -0,0 +1,71 @@
+import { httpGet$, httpPost$ } from "./http"
+import { of } from 'rxjs';
+import { map, switchMap, share, finalize } from 'rxjs/operators';
+
+const BASE_URL = `https://api.spotify.com/v1/browse`;
+const CLIENT_ID = process.env.REACT_APP_SPOTIFY_CLIENT_ID;
+const CLIENT_SECRET = process.env.REACT_APP_SPOTIFY_CLIENT_SECRET;
+
+let tokenExpireAt = null;
+let currentToken = null;
+let authTokenSource$ = null;
+
+const getAuthToken$ = () => {
+  const body = 'grant_type=client_credentials';
+
+  const headers = {
+    'Authorization': 'Basic ' + (new Buffer.from(CLIENT_ID + ':' + CLIENT_SECRET).toString('base64')),
+    'Content-Type': 'application/x-www-form-urlencoded',
+  }
+  return httpPost$('https://accounts.spotify.com/api/token', body, headers);
+}
+
+const getAccessToken$ = () => {
+  if (!currentToken || tokenExpireAt < Date.now()) {
+    if (!authTokenSource$) {
+      authTokenSource$ = getAuthToken$().pipe(
+        map(({ access_token, expires_in }) => {
+          tokenExpireAt = Date.now() + expires_in * 1000;
+          currentToken = access_token
+          return currentToken;
+        }),
+        finalize(() => authTokenSource$ = null),
+        share()
+      )
+    }
```

Comment:
Above is a singleton design that can make sure concurrent Spotify requests can maximumly trigger only one request to get the access token.